parlai/core/torch_generator_agent.py

```diff
@@ -984,3 +1012,66 @@ def select_paths(self, logprobs, prior_scores):
         tok_ids = best_idxs % voc_size
 
         return (hyp_ids, tok_ids, best_scores)
+
+
+class TopKSampling(TreeSearch):
+    """
+    Top-K sampling (Fan et al., 2018).
+
+    Samples from a truncated distribution where only the most probable K words
+    are considered at each time.
+
+    Typical values of k are 2, 10, 50.
+
+    See https://arxiv.org/abs/1805.04833 for details.
+    """
+
+    def __init__(self, k, *args, **kwargs):
+        self.k = k
+        super().__init__(*args, **kwargs)
```

Comment:
Is setting `self.k` done before calling base class `__init__` for some reason? I've only seen this flipped around where first the base class is initialized and then local initialization is done. I think this opens up the possibility of the base class setting `self.k` as well and overriding what you want to do here (although I know it doesn't here, just commenting on the pattern).